File: Leetcode/Algorithm/python/Longest Consecutive Sequence.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def longestConsecutive(self, num):
        s = set(num)
        visited = set()
        ans = 0
        for i in num:
            l = i - 1
            r = i + 1
            if i in visited:
                continue
            visited.add(i)
            v = 1
            while l in s:
                v += 1
                visited.add(l)
                l -= 1
            while r in s:
                v += 1
                visited.add(r)
                r += 1
            ans = max(ans, v)
        return ans
                

File: LintCode/cpp/Longest Consecutive Sequence.cc
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return an integer
     */
    int longestConsecutive(vector<int> &num) {
        for (auto i: num) {
            father[i] = i;
        }
        
        for (auto i: num) {
            int a = i - 1;
            int b = i + 1;
            
            if (father.find(a) != father.end()) {
                father[find_father(a)] = find_father(i);
            }
            
            if (father.find(b) != father.end()) {
                father[find_father(b)] = find_father(i);
            }
        }
        
        int ans = 0;
        unordered_map<int, int> cnt;
        for (auto p: father) {
            cnt[p.second]++;
            ans = max(ans, cnt[p.second]);
        }
        return ans;
    }
private:
    int find_father(int x) {
        if (father[x] == x) {
            return x;
        }
        return father[x] = find_father(father[x]);
    }
private:
    unordered_map<int, int> father;
};
